PSPs/drafts/psp-6.md

```diff
@@ -0,0 +1,2903 @@
+
+All parameters in the specified JSON-RPC methods are **REQUIRED**, unless explicitly mentioned otherwise (**OPTIONAL**). Return values which are OPTIONAL indicate the possibility of `null` being returned.
+
+Additionally, the JSON-RPC API should be made available over the [WebSocket Protocol](https://tools.ietf.org/html/rfc6455).
+
+### 1.3.1. Safety
+
+Exposing RPC calls to the public may open up a huge surface of attacks, such as denial or service attacks, and has to be carefully considered. There are quite a few RPC methods that should be never (or rarely) directly exposed. Those methods are considered **UNSAFE** and are explicitly highlighted.
+
+The implementation of this specification must disable the JSON API by default and should offer four different flags (for the RPC/TCP and the RPC/WebSocket interface each) in order to enable specific parts of the API. Two flags for exposing safe and unsafe PRC to localhost only, and two flags for exposing safe and unsafe RPC to all interfaces.
+
+| API / Network | Localhost                 | All interfaces             |
+| ------------- | ------------------------- | -------------------------- |
+| Safe API      | `--local-{rpc/ws}`        | `--public-{rpc/ws}`        |
+| Unsafe API    | `--local-unsafe-{rpc/ws}` | `--public-unsafe-{rpc/ws}` |
+
+**Warning**: Enabling `--public-unsafe-{rpc|ws}` is dangerous and should be avoided.
+
+### 1.3.2. Errors
+
+Each namespace specifies RPC error code ranges of possible error types that can be returned. Known error types are defined in this document and some APIs may return undefined error messages. Defined error messages are subject to expansion and changes.
+
+### 1.3.3. Common types
+
+This document references commonly used types, represented in capital letters, as defined in the subsections below.
+
+#### 1.3.3.1. STRING
+
+A sequence of characters: `"value"`
+
+#### 1.3.3.2. U32 / U64
+
+Unsigned integers.
+
+* `U32` - A 32-byte unsigned integer.
+* `U64` - A 64-byte unsigned integer.
+
+#### 1.3.3.3. BOOL
+
+A boolean type, which can either be `true` or `false`.
+
+#### 1.3.3.4. MAP
+
+A key/value object.
+
+```json
+{
+    "key": "value1",
+    "key2": "value2"
+}
+```
+
+#### 1.3.3.5. ARRAY
+
+An object containing a varying amount of items.
+
+```json
+[
+    "item1",
+    "item2"
+]
+```
+
+#### 1.3.3.6. HASH
+
+A string which represents a 32-byte, hexadecimal-encoded Blake2 hash. Contains a `0x` prefix to indicate the hexadecimal encoding.
+
+#### 1.3.3.7. HEX
+
+A string of varying size representing hexadecimal-encoded data. Contains a `0x` prefix to indicate the hexadecimal encoding.
```

Comment:
Would it be better to put this before `HASH` since all hashes are hex? Also, consider explicitly enumerating the valid characters, e.g. `[a-fA-F0-9]`

```suggestion
#### 1.3.3.7. HEX

A variable-size string that represents hexadecimal-encoded data. Must begin with the characters `0x` to indicate a hexadecimal encoding, e.g. `0x1337`.
```

Reply:
> Would it be better to put this before HASH since all hashes are hex?

Done!

> Also, consider explicitly enumerating the valid characters, e.g. [a-fA-F0-9]

Given that it explicitly states "hexadecimal-encoded", I guess this is somewhat redundant. But I added it anyway.